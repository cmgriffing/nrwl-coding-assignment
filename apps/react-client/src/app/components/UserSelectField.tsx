import React from 'react';
import { User } from '@acme/shared-models';

interface UserSelectFieldProps {
  onChange: (userId: number | null) => Promise<void> | void;
  users: User[];
  selectedUserId: number | null;
}

const UNASSIGNED = 'Unassigned';

// Not entirely necessary to extract this to a component,
export function UserSelectField({
  onChange,
  users,
  selectedUserId,
}: UserSelectFieldProps) {
  return (
    <select
      value={selectedUserId || undefined}
      onChange={(event) => {
        let newAssigneeId = null;
        if (event.currentTarget.value !== UNASSIGNED) {
          newAssigneeId = Number(event.currentTarget.value);
        }
        onChange(newAssigneeId);
      }}
    >
      <option>{UNASSIGNED}</option>
      {users.map(({ id, name }) => (
        <option key={id} value={id}>
          {name}
        </option>
      ))}
    </select>
  );
}
