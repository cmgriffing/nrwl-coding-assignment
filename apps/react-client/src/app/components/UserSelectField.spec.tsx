import React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import { User } from '@acme/shared-models';

import { UserSelectField } from './UserSelectField';

const emptyUsersList: User[] = [];
const usersList: User[] = [
  { id: 1, name: 'Alice' },
  { id: 2, name: 'Bob' },
  { id: 3, name: 'Chris' },
  { id: 4, name: 'Daisy' },
  { id: 5, name: 'Ed' },
];

it('renders an empty list', () => {
  const { asFragment, getAllByRole } = render(
    <UserSelectField
      users={emptyUsersList}
      selectedUserId={null}
      onChange={() => {}}
    />
  );

  console.log(asFragment());
  expect(asFragment).toMatchSnapshot();
  expect(getAllByRole('option').length).toEqual(1);
});

it('renders a list without selected user', () => {
  const { asFragment, getAllByRole } = render(
    <UserSelectField
      users={usersList}
      selectedUserId={null}
      onChange={() => {}}
    />
  );
  expect(asFragment()).toMatchSnapshot();
  expect(getAllByRole('option').length).toEqual(usersList.length + 1);
});

it('renders a list with selected user', () => {
  const { asFragment, getAllByRole } = render(
    <UserSelectField
      users={usersList}
      selectedUserId={usersList[2].id}
      onChange={() => {}}
    />
  );
  expect(asFragment()).toMatchSnapshot();
  expect(getAllByRole('option').length).toEqual(usersList.length + 1);
});
