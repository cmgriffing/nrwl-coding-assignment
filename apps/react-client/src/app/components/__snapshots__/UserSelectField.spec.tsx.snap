// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a list with selected user 1`] = `
<DocumentFragment>
  <select>
    <option>
      Unassigned
    </option>
    <option
      value="1"
    >
      Alice
    </option>
    <option
      value="2"
    >
      Bob
    </option>
    <option
      value="3"
    >
      Chris
    </option>
    <option
      value="4"
    >
      Daisy
    </option>
    <option
      value="5"
    >
      Ed
    </option>
  </select>
</DocumentFragment>
`;

exports[`renders a list without selected user 1`] = `
<DocumentFragment>
  <select>
    <option>
      Unassigned
    </option>
    <option
      value="1"
    >
      Alice
    </option>
    <option
      value="2"
    >
      Bob
    </option>
    <option
      value="3"
    >
      Chris
    </option>
    <option
      value="4"
    >
      Daisy
    </option>
    <option
      value="5"
    >
      Ed
    </option>
  </select>
</DocumentFragment>
`;

exports[`renders an empty list 1`] = `[Function]`;
